---
title: "SARS-CoV-2 QC report"
nav: "Quality control"
author: "BIMSB Bioinformatics Platform"
date: '`r format(as.POSIXct(if ("" != Sys.getenv("SOURCE_DATE_EPOCH")) { as.numeric(Sys.getenv("SOURCE_DATE_EPOCH")) } else { Sys.time() }, origin="1970-01-01"), "%Y-%m-%d %H:%M:%S")`'
params:
  sample_name: ""
  coverage_file: ""
  logo: ""
  coverage_table_outfile: ""
  multiqc_ran: false
  multiqc_rel_path: ""
---

<style>
.dropdown-menu {
  max-height: 200px;
  overflow-y: scroll;}
</style>

<div id="logo" align="top">
`r knitr::include_graphics(params$logo)`
</div>


<sample>`r params$sample_name`</sample>

```{r setup, include = FALSE, message = FALSE, warning = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE
)

knitr::knit_hooks$set(
  warning = function(x, options) {
    paste('\n\n<div class="alert alert-warning">',
      gsub("##", "\n", gsub("^##\ Warning:", "**Warning**", x)),
      "</div>",
      sep = "\n"
    )
  }
)

library(dplyr)
library(DT)
library(magrittr)
library(base64url)
library(stringr)
library(data.table)

multiqc_ran <- as.logical(params$multiqc_ran)
```

This report provides an overview over general quality control measures of the
sample and provides the summarized data as downloadable tables. For detailed
information about the methods please consult the
[documentation](http://bioinformatics.mdc-berlin.de/pigx_docs/pigx-sars-cov-2.html#output-description).

```{r reading_data, include = FALSE}
# FIXME Rename these col names sensibly upstream
coverage_df <- fread(params$coverage_file)
```

# Coverage reports
Here, the information about read coverage of the reference genome as well as
coverage of the mutation locations that were provided are listed. From those
values, the overall quality of the sample can be assessed and whether or not its
analysis results should and will be used for further trend-analysis and
visualization.

## Mutation locations

In order to assess variant calling based on mutations of interest it is crucial
that its genome locations are covered by read alignment. The table below shows
the proportion of covered mutation locations, how many mutations are tracked in
total, how many of their locations are covered and which locations have no read
coverage at all.
 
```{r amplicon_coverage, echo = FALSE}
mutation_coverage_cols <- c(
  "n_loc_total",
  "n_loc_covered",
  "n_loc_no_cov"
)

mut_cov_df <- coverage_df %>%
  select(all_of(mutation_coverage_cols)) %>%
  rename(c(
    "Total number of tracked mutations" = mutation_coverage_cols[1],
    "Number of mutations covered" = mutation_coverage_cols[2],
    "Mutation locations (NT) not covered" = mutation_coverage_cols[3]
  ))


datatable(
  t(mut_cov_df) %>% set_colnames("Value"),
  options = list(searching = FALSE, paging = FALSE))
```

## Reference Genome

This table shows the total number of raw reads, number of aligned reads, the
mean coverage depth and percentage of the reference genome covered by
pre-processed reads.

```{r read_coverage, echo = FALSE}
read_coverage_cols <- c(
  "numreads",
  "coverage",
  "meandepth"
)

read_cov_df <- coverage_df %>%
  select(all_of(read_coverage_cols)) %>%
  rename(c(
    "Number of aligned reads" = read_coverage_cols[1],
    "Total ref. genome coverage" = read_coverage_cols[2],
    "Mean read depth" = read_coverage_cols[3])) %>%
  mutate(
    `Total ref. genome coverage` = paste(round(as.integer(
      `Total ref. genome coverage`
    ), 1), "%"),
    `Mean read depth` = round(as.integer(
      `Mean read depth`
    )))

datatable(t(read_cov_df) %>% set_colnames("Value"))
```

# Read processing

To improve alignment rates and mutation calling the pipeline contains trimming
steps for primer, read-quality and adapter trimming. Parameter for read
processing can be provided by 
[the pipelineâ€™s settings file](link/to/etc/settings_file). For detailed method 
information please
consult the 
[documentation](http://bioinformatics.mdc-berlin.de/pigx_docs/pigx-sars-cov-2.html#introduction).

## MultiQC 

``` {r multiqc_warning, echo = FALSE, warning = TRUE}
if (!multiqc_ran) {
  warning(
    paste("No MultiQC reports were generated for this run, this is most ",
          "likely due to the provided input files not being fastq files.",
          sep = ""
    )
  )
}
```


```{r multiqc_text, echo = FALSE, results='asis'}
if (multiqc_ran) {
  cat(paste(
    "In order to be able to assess and compare the read quality after each",
    "step multiple quality reports are generated which are presented below in",
    "a summarized format:\n\n",
    "Download:\n",
    "[MultiQC report of raw and processed data](", params$multiqc_rel_path, ")"
  ))
}
```

## Trimming

To improve alignment rates and mutation calling, proper trimming of reads should
be considered e.g. based on the provided fastqc-reports.

# Download Stats csv

```{r data_download, include = FALSE}
coverage <- readLines(params$coverage_file) %>%
  paste0(collapse = "\n") %>%
  base64_urlencode()
```

Coverage statistics:  
[Download Coverage_stats.csv](`r sprintf('data:text/csv;base64,%s', coverage)`)
