---
title: "SARS-CoV-2 Mutations and Lineage report"
nav: "Variant report"
author: "BIMSB Bioinformatics Platform"
date: '`r format(as.POSIXct(if ("" != Sys.getenv("SOURCE_DATE_EPOCH")) { as.numeric(Sys.getenv("SOURCE_DATE_EPOCH")) } else { Sys.time() }, origin="1970-01-01"), "%Y-%m-%d %H:%M:%S")`'
params:
  sample_name: ""
  sigmut_file: ""
  non_sigmut_file: ""
  variant_abundance_file: ""
  snv_file: ""
  vep_file: ""
  vep_file_raw: ""
  vep_transcript_distance: ""
  vep_species: ""
  logo: ""
---

<style>
.dropdown-menu {
  max-height: 200px;
  overflow-y: scroll;}
</style>

<div id="logo" align="top">
`r knitr::include_graphics(params$logo)`
</div>

<sample>`r params$sample_name`</sample>

```{r setup, include = FALSE, warning = FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

library(MASS)
library(knitr)
library(dplyr)
library(data.table)
library(ggplot2)
library(DT)
library(base64url)
library(stringr)
```

This lineage analysis report provides an overview over all mutations found in 
the sample and the deconvoluted lineage abundances. The tables below display the
summarized results from single-nucleotide-variant (SNV) calling and their 
translation into protein mutations. A bar chart visualizes the proportions of 
lineages that are derived by deconvolution of the set of mutations that is 
matching the provided list of lineage - characterizing mutations (signature 
mutations). For detailed method description please consult the
[documentation](http://bioinformatics.mdc-berlin.de/pigx_docs/pigx-sars-cov-2.html#output-description)
*Note: Deconvolution results are biased by the number of reads and by the number
of signature mutations that were found for this sample.  Check with the QC 
report to be able to assess how trustworthy the results are.*


# Mutation analysis

The following tables combine the results from the variant calling which are 
SNVs, their frequency and read depth. They get translated to get the info 
about the resulting protein mutation and on which gene they are located. If the
SNV matches a provided signature mutation it is marked which of the provided
lineages they characterize. 

Table 1: SNVs from sample matching signature mutations

```{r sig_table, results = 'asis'}
match.df <- fread(params$sigmut_file)

# Display the filtered signature mutations and the variants accordingly in
#  scrollable tables
SignatureTable <- data.frame(
  "SNV" = match.df$mut_str,
  "gene" = match.df$gene_name,
  "protein_mutation" = match.df$aa_str,
  "variant" = match.df$variant, # TODO: more nice writing of it e.g B1.1.7
  "frequency" = match.df$freq, # TODO: more clear name stating which frequency
  "depth" = match.df$dep
) # TODO: more clear title stating which depth

DT::datatable(SignatureTable,
  extensions = c("FixedColumns", "Scroller"),
  options = list(
    fixedColumns = TRUE,
    scrollY = 180,
    scrollX = TRUE,
    scroller = TRUE,
    dom = "Bfrtip"
  ),
  filter = "bottom"
)
```

Table 2: SNVs from sample not matching any signature mutations

```{r non_sig_table, results = 'asis'}
nomatch.df <- fread(params$non_sigmut_file)

# Display the filtered signature mutations and the variants accordingly in
# scrollable tables
NoSignatureTable <- data.frame(
  "SNV" = nomatch.df$mut_str,
  "gene" = nomatch.df$gene_name,
  "protein_mutation" = nomatch.df$aa_str,
  "frequency" = nomatch.df$freq, # TODO more clear name stating which frequency
  "depth" = nomatch.df$dep
) # TODO more clear title stating which depth

DT::datatable(NoSignatureTable,
  extensions = c("FixedColumns", "Scroller"),
  options = list(
    fixedColumns = TRUE,
    scrollY = 400,
    scrollX = TRUE,
    scroller = TRUE,
    dom = "Bfrtip"
  ),
  filter = "bottom"
)
```

# Lineage analysis

Lineage proportion values are derived by deconvolution of the tracked signature
mutations. Deconvolution is done by using a robust regression model. It aims to
predict the individual contributions of the lineages to the observed mutation
frequencies in the bulk wastewater sample. For detailed method description,
please consult the
[documentation](https://bioinformatics.mdc-berlin.de/pigx_docs/pigx-sars-cov-2.html#output-description)

```{r echo = FALSE}
# get  NT mutations only, input for the signature matrix
mutations.vector <- match.df$gene_mut_collapsed

df <- fread(params$variants_file)

# only execute the deconvolution when at least one signature mutation was found
executeDeconvolution <- nrow(df) > 0
```

```{asis, echo=executeDeconvolution}
The following bar plot shows the proportion of each lineage in the sample as
predicted by the deconvolution model. For the case of multiple lineages being
indistinguishable because of shared signature mutations, the deconvolution
result for this group is distributed equally across each grouped lineage and
they all get the same value.
```

```{r plot, eval = executeDeconvolution}
Variant_colors <- c(
  "#8a3874",
  "#4d3da1",
  "#62c742",
  "#13317a",
  "#86e5a3",
  "#6f69e5",
  "#4cdab2",
  "#999ce4",
  "#2ce6f3",
  "#6638c6",
  "#005f29",
  "#5d73cc",
  "#616524",
  "#a8a041",
  "#e9f600",
  "#3eaca8",
  "#62a273",
  "#4ea5d0",
  "#655c95",
  "#67a04e"
)
# TODO this should be put into a separate color palette function in a separate file

font_family <- "DejaVu Sans"

ggplot(data = df, aes(variant, abundance,
  fill = variant
)) +
  geom_bar(stat = "identity") +
  labs(
    title = "Proportion of variants of concern inferred from deconvolution",
    x = "Variants",
    y = "Percentage"
  ) +
  scale_x_discrete(limits = df$variant) +
  scale_y_continuous(
    limits = c(0, 1), breaks = seq(0, 1, 0.1),
    labels = scales::percent_format(accuracy = 1)
  ) +
  scale_colour_manual(values = Variant_colors) +
  # geom_text ignores the theme's default font family, so we
  # need to provide it here.
  geom_text(aes(
    family = font_family,
    label = sprintf("%0.3f", round(abundance, digits = 3))
  ),
  position = position_dodge(width = 0.9),
  vjust = -0.25
  ) +
  scale_fill_discrete(
    name = "Variants",
    breaks = df$variant,
    labels = df$variant
  ) +
  theme_gray(base_family = font_family)
```

# Analysis Parameter

```{r extract_analysis_parameters}
# the info is contained in the header, so we only need the first couple of lines
vep_assembly <- readLines(params$vep_file_raw, n = 500) %>%
  paste(collapse = "") %>%
  str_extract("(?<=## assembly version )[[:alnum:]]+")

trans_dist <- params$vep_transcript_distance
```

## LoFreq filter
| LoFreq is used with default filter settings:       |                   |
| :------------------------------------------------- | :---------------- |
| minimum coverage                                   | 10                |
| Strand-Bias Multiple Testing Correction, fdr corr. | pvalue > 0.001000 |
| Minimum SNV Quality (Phred)                        | 63                |
| Minimum Indel Quality (Phred)                      | 47                |

## VEP settings  
* Species: `r params$vep_species` 
* Assembly: `r vep_assembly`  

|                                                   |                  |
| :------------------------------------------------ | :--------------- |
| Matched Annotations from NCBI and EMBL-EBI (MANE) | yes              |
| Restrict results by severity of consequence       | no               |
| upstream/downstream distance                      | `r trans_dist`bp |


# Raw data 

```{r download_raw, include = FALSE}
## ------------------------------------------------
# make raw/input data downloadable
snvs <- readLines(params$snv_file) %>%
  paste0(collapse = "\n") %>%
  base64_urlencode()

vep <- readLines(params$vep_file) %>%
  paste0(collapse = "\n") %>%
  base64_urlencode()
```


Called SNVs:  
[Download SNVs.csv](`r sprintf('data:text/csv;base64,%s', snvs)`)

Variant effect prediction (*Ensemble VEP output*):  
[Download VEP_out.csv](`r sprintf('data:text/csv;base64,%s', vep)`)
